public/app/features/scopes/selector/ScopesSelectorService.ts

```diff
@@ -1,260 +1,257 @@
+   * Selecting a scope means we add it to a temporary list of scopes that are waiting to be applied. We make sure
+   * that the selection makes sense (like not allowing selection from multiple categories) and prefetch the scope.
+   * @param scopeNodeId
    */
-  public toggleNodeSelect = (node: ToggleNode) => {
-    if ('scopeName' in node) {
-      // This is for a case where we don't have a path yet. For example on init we get the selected from url, but
-      // just the names. If we want to deselect them without knowing where in the tree they are we can just pass the
-      // name.
-
-      const newTreeScopes = this.state.treeScopes.filter((s) => s.scopeName !== node.scopeName);
-      if (newTreeScopes.length !== this.state.treeScopes.length) {
-        this.updateState({ treeScopes: newTreeScopes });
-        return;
-      }
-    }
+  public selectScope = async (scopeNodeId: string) => {
+    let scopeNode = this.state.nodes[scopeNodeId];
 
-    if (!node.path) {
-      console.warn('Node cannot be selected without both path and name', node);
-      return;
+    if (!isNodeSelectable(scopeNode)) {
+      throw new Error(`Trying to select node with id ${scopeNodeId} that is not selectable`);
     }
 
-    let treeScopes = [...this.state.treeScopes];
-    const parentNode = getNodesAtPath(this.state.nodes[''], node.path.slice(1, -1));
-
-    if (!parentNode) {
-      // Either the path is wrong or we don't have the nodes loaded yet. So let's check the selected tree nodes if we
-      // can remove something based on scope name.
-      const scopeName = node.path.at(-1);
-      const newTreeScopes = treeScopes.filter((s) => s.scopeName !== scopeName);
-      if (newTreeScopes.length !== treeScopes.length) {
-        this.updateState({ treeScopes: newTreeScopes });
-      } else {
-        console.warn('No node found for path:', node.path);
-      }
-      return;
+    if (!scopeNode.spec.linkId) {
+      throw new Error(`Trying to select node id ${scopeNodeId} that does not have a linkId`);
     }
 
-    const nodeName = node.path[node.path.length - 1];
-    const { linkId, title } = parentNode.nodes[nodeName];
-
-    const selectedIdx = treeScopes.findIndex(({ scopeName }) => scopeName === linkId);
-
-    if (selectedIdx === -1) {
-      // We are selecting a new node.
-
-      // We prefetch the scope when clicking on it. This will mean that once the selection is applied in closeAndApply()
-      // we already have all the scopes in cache and don't need to fetch all of them again is multiple requests.
-      this.apiClient.fetchScope(linkId!);
+    // We prefetch the scope metadata to make sure we have it cached before we apply the scope.
+    this.apiClient.fetchScope(scopeNode.spec.linkId).then((scope) => {
+      // We don't need to wait for the update here, so we can use then instead of await.
+      if (scope) {
+        this.updateState({ scopes: { ...this.state.scopes, [scope.metadata.name]: scope } });
+      }
+    });
```

Comment:
this looks like an api call, should there not be a promise (Await) / error handling somewhere here? or is that handled inside fetchScope ?

Reply:
I think ideally this would be something like a `useGetScope(scopeId)` hook that will return the data either by calling the API or from cache, or fallback data, without having to keep the map in the service, or figuring out when to load what but this PR is too big already for that IMHO.